dialectic-compiler/src/lib.rs

```diff
@@ -0,0 +1,909 @@
+use {
+    lazy_static::lazy_static,
+    proc_macro2::{Span, TokenStream},
+    quote::{quote, ToTokens},
+    std::{fmt, ops, rc::Rc},
+    syn::{spanned::Spanned, Error, Ident, Type},
+    thiserror::Error,
+    thunderdome::{Arena, Index},
+};
+
+pub mod parse;
+
+#[derive(Error, Debug, Clone)]
+pub enum CompileError {
+    #[error("label name `'{0}` shadows a label name that is already in scope")]
+    ShadowedLabel(String),
+    #[error("undeclared label `'{0}`")]
+    UndeclaredLabel(String),
+    #[error("cannot `continue` outside of a loop")]
+    ContinueOutsideLoop,
+    #[error("cannot `break` outside of a loop")]
+    BreakOutsideLoop,
+}
+
+#[derive(Debug, Clone)]
+pub struct SpannedCompileError {
+    span: Span,
+    kind: CompileError,
+}
+
+/// A shim for parsing the root level of a macro invocation, so
+/// that a session type may be written as a block w/o an extra
+/// layer of braces.
+#[derive(Debug, Clone)]
+pub struct Invocation {
+    pub syntax: SyntaxNode,
+}
+
+#[derive(Debug, Clone)]
+pub struct SyntaxNode {
+    pub expr: Syntax,
+    pub span: Span,
+}
+
+#[derive(Debug, Clone)]
+pub enum Syntax {
+    Recv(Type),
+    Send(Type),
+    Call(Box<SyntaxNode>),
+    Choose(Vec<SyntaxNode>),
+    Offer(Vec<SyntaxNode>),
+    Split(Box<SyntaxNode>, Box<SyntaxNode>),
+    Loop(Option<String>, Box<SyntaxNode>),
+    Break(Option<String>),
+    Continue(Option<String>),
+    Block(Vec<SyntaxNode>),
+    Type(Type),
+}
+
+#[derive(Debug, Clone)]
+struct CfgNode {
+    expr: Ir,
+    next: Option<Index>,
+    span: Span,
+}
+
+#[derive(Debug, Clone)]
+enum Ir {
+    Done,
+    Recv(Type),
+    Send(Type),
+    Call(Index),
+    Choose(Vec<Index>),
+    Offer(Vec<Index>),
+    Split(Index, Index),
+    Loop(Option<String>, Index),
+    IndexedBreak(usize),
+    IndexedContinue(usize),
+    LabeledBreak(String),
+    LabeledContinue(String),
+    Type(Type),
+    Error,
```

Comment:
Why don't we carry error information to propagate in the `Ir`? I see that we collect errors in an associated `Vec` during compilation; why not embed errors in the generated `Ir`, continue compilation to the `Target`, embed errors there too, and only during final code generation extract all errors? This would maximize our ability to detect as many errors in one go as possible, it seems to me.

Reply:
Originally, I did have that. I ran into some problems which I've since forgotten and swapped to this scheme. I finished this late on Friday so I'll take a look at making that better today; I don't like the errors-collected-as-side-effects scheme either.